PovioKit/Classes/ShareKit/Asset Library/SKAssetLibraryProvider.swift

```diff
@@ -0,0 +1,152 @@
+//
+//  SKAssetLibraryProvider.swift
+//  PovioKit
+//
+//  Created by Toni Kocjan on 37/12/2018.
+//  Copyright © 2018 Povio Labs. All rights reserved.
+//
+
+import Photos
+
+protocol SKAssetLibraryTemplate {
+  func authorizeIfNecessary(success: (() -> Void)?, failure: (() -> Void)?)
+  func findOrCreateAlbum(name: String, completion: @escaping (PHAssetCollection?) -> Void)
+  func createAlbum(name: String, completion: @escaping (PHAssetCollection?) -> Void)
+  func saveImage(image: UIImage, album: PHAssetCollection, completion: @escaping (PHAsset?) -> Void)
+  func saveVideo(filePath path: String, album: PHAssetCollection, completion: @escaping (PHAsset?) -> Void)
+  func urlForAsset(_ asset: PHAsset, completion: @escaping (URL?) -> Void)
+  func clearAlbum(_ album: PHAssetCollection, completion: ((Bool) -> Void)?)
+}
+
+class SKAssetLibraryProvider: SKAssetLibraryTemplate {
+  private enum AssetType: Hashable {
+    case image(UIImage)
+    case video(String)
+  }
+  private var assetCache = [AssetType: PHAsset]()
+  
+  func authorizeIfNecessary(success: (() -> Void)?, failure: (() -> Void)?) {
+    guard PHPhotoLibrary.authorizationStatus() != .authorized else {
+      success?()
+      return
+    }
+    
+    PHPhotoLibrary.requestAuthorization { status in
+      status == .authorized ? success?() : failure?()
+    }
+  }
+  
+  func findOrCreateAlbum(name: String, completion: @escaping (PHAssetCollection?) -> Void) {
+    let fetchOptions = PHFetchOptions()
+    fetchOptions.predicate = NSPredicate(format: "title = %@", name)
+    guard let photoAlbum = PHAssetCollection.fetchAssetCollections(with: .album, subtype: .any, options: fetchOptions).firstObject else {
+      return createAlbum(name: name, completion: completion)
+    }
+    return completion(photoAlbum)
+  }
+  
+  func createAlbum(name: String, completion: @escaping (PHAssetCollection?) -> Void) {
+    var albumPlaceholder: PHObjectPlaceholder?
+    PHPhotoLibrary.shared().performChanges({
+      let createAlbumRequest = PHAssetCollectionChangeRequest.creationRequestForAssetCollection(withTitle: name)
+      albumPlaceholder = createAlbumRequest.placeholderForCreatedAssetCollection
+    }, completionHandler: { success, error in
+      guard success else {
```

Comment:
We can just join these 3 guards together since we're not doing any specific fallbacks/error handling